internal/appsec/waf/waf.go

```diff
@@ -624,6 +751,18 @@ func decNbLiveCObjects() {
 	atomic.AddUint64(&nbLiveCObjects, ^uint64(0))
 }
 
+// gostring returns the Go version of the C string `str`, copying at most `len` bytes from the original string.
+func gostring(str *C.char, len C.uint64_t) (string, error) {
+	if str == nil {
+		return "", ErrInvalidArgument
+	}
+	goLen := C.int(len)
+	if C.uint64_t(goLen) != len {
+		return "", ErrInvalidArgument
```

Comment:
Do you know any bitwise operation trick to safely convert it to `int` when it's safe to do so?

Can we do something like:

```go
if C.int(len) <= 0 { /* overflow? */ }
```

Worth having some unit tests for this condition.

Reply:
We could check if the upper half of the uint64_t is zeroed out and only cast in this case, and report an error otherwise.
Something like `(len >> 32) & 0xFFFFFFFF == 0`

Reply:
As discussed, the current implementation is good enough and this case should never happen, so it is correctly caught at the moment.

As for WASM, which will be 32 bits, it won't use that function because it doesn't rely on CGO anyway.